File: algorithms/swift/28-implement_strstr.swift
// https://leetcode.com/problems/implement-strstr/submissions/

class Solution {
    
    // KMP Solution
    func strStr(_ haystack: String, _ needle: String) -> Int {
        if haystack == needle || needle.isEmpty { return 0 }
        if haystack.count <= needle.count || haystack.isEmpty { return -1 }
        
        let pattern = Array(needle)
        let text = Array(haystack)
        
        // Preprocessing - compute prefix
        var pi = [Int](repeating:0, count:pattern.count)
        var p = 0
        for i in 1..<pattern.count {
            while p > 0 && pattern[i] != pattern[p] {
                p = pi[p-1]
            }
            if pattern[i] == pattern[p] {
                p += 1
            } 
            pi[i] = p
        }
        // Matching
        var q = 0
        for i in 0..<text.count {
            while (q > 0 && pattern[q] != text[i]) {
                q = pi[q-1]
            }
            if (pattern[q] == text[i]) {
                q += 1
                if (q == pattern.count) { 
                    return i - pattern.count + 1
                }
            }
        }
        return -1
    }
}

File: algorithms/python/28-implement-strstr.py
# https://leetcode.com/problems/implement-strstr/submissions/

class Solution(object):
    
    # Naïve solution
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if needle == "": return 0
        if len(haystack) < len(needle): return -1
        
        for i in range(0,len(haystack)-len(needle)+1):
            if haystack[i: i+len(needle)] == needle:
                return i
        return -1
            
    # KMP Solution
    def strStr2(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if needle == "": return 0
        if len(haystack) < len(needle): return -1
        
        # Preprocessing
        pi = [0]
        i = 0
        for j in range(1, len(needle)):
            while i > 0 and needle[i] != needle[j]:
                i = pi[i-1]
            if needle[j] == needle[i]:
                i += 1
            pi.append(i)
        
        # Matching
        q = 0
        for k in range(0, len(haystack)):
            while q > 0 and haystack[k] != needle[q]:
                q = pi[q-1]
            if needle[q] == haystack[k]:
                q += 1
                if (q == len(needle)):
                    return k - len(needle) + 1
        return -1